### DISCUSSION TITLE
The semantics of `insomnia.execution.setNextRequest()`, request looping

### DISCUSSION THREAD

**Discussion by janspudich** [Kong/insomnia#8250] [Help] [answered]
Dear members of the Insomnia community,

please help me solve the following problem related to a flow of request calls when running a collection/folder in Insomnia.

Let's assume that there are three requests:
- `Request A`: Discover all resources, say projects
- `Request B`: Fetch an additional piece of information for a single resource, say project risks
- `Request C`: Fetch an additional piece of information for a single resource, say project issues

Now, the goal is to run the requests so that we discover all projects (using `Request A`) and for each project returned by `Request A` we fetch both risks (using `Request B`) and issues (using `Request C`). This implies a loop of the requests B and C. The number of iterations in the loop is determined by the response to a call of `Request A`.

These are my observations which I made while trying to achieve the goal (using Insomnia v10.2.0 on macOS):

1. When I use `setNextRequest('Request B')` in `After-response` of `Reqeust C`, it does not have any impact on the flow.
3. When I use `setNextRequest('Request C')` in `After-response` of `Reqeust C`, it loops over the request C (I can loop, at least).
4. Not strictly related to the goal but I mention it for the sake of completeness; when I use `setNextRequest('Request X')` in `Pre-request` of `Request B`, it does not prevent Insomnia from calling `Request B` first before moving to `Request X` (which I dare to say is not perfectly in line with the [Dynamically set the next request](https://docs.insomnia.rest/insomnia/collection-runner#dynamically-set-the-next-request) paragraph in the Insomnia documentation).
5. #7628 did not help.

Having written all that, can someone give me a guidance to achieve the goal and help me close the gap I have in understanding how `setNextRequest()` works in Insomnia? Is the purpose of `setNextRequest()` to let a user control a flow of request calls within a collection/folder run or not?

Cheers,
Jan

**Comment by subnetmarco** [comment]
@ihexxa

**Comment by ihexxa** [comment, marked as answer]
Hello @janspudich,

In short, `setNextRequest` has following behaviors:
- It can jump to any request which follows the current request, or itself (the case 2 you mentioned)
- It can not go back to any previous request. (the case 1 you mentioned)
- The request will still be sent if it is called in pre-request script (the case 3)

For your case, I assume it is allowed to:
1. Collect all risks at first (send Request B for all resources in a loop)
2. Collect all issues secondly (send Request C for all resources in another loop)

You could try something like:
- Start runner with 3 requests:
  - Req A 
  - Req B
  - Req C
- When got resources from `Req A`, in script, set all resources as an array by `insomnia.variables.set('resources', [...])`.
- When calling `Req B`, iterate all resources saved and prepare the pre-script like:
```javascript
// get the next resource to send req b
const resources = insomnia.variables.get('resources');
const idx = insomnia.variables.get('idx') || 0;
const currentResource = resources[idx];

// do something you want

if (idx+1 < resources.length) { // redirect to self to collect risks
  insomnia.variables.set('idx', idx+1);
  insomnia.execution.setNextRequest('Request B');
} else { // start collect issues
  insomnia.variables.set('idx', 0);
  insomnia.execution.setNextRequest('Request C');
}
```
- When calling `Req C`, do something similar to above.

**Reply by janspudich** [reply]
Hi @ihexxa,

first of all, let me thank you for your comprehensive response, incl. the clarification of the `setNextRequest` semantics.

Regarding your proposal as to how the problem declared in my original post can be solved: honestly, the approach that you are suggesting was my plan B if it turned out that my understanding of `setNextRequest` was complete. I was hoping that there was something which I was missing and which would allow me to iterate over a sequence of requests. Apparently, that is not the case and thus I will have to do my plan B, as you suggest.

The only thing is that the use-case as presented above is a distilled (and simplified) version of the real use-case and the amount of overhead to manage the loop for each request separately is so high that it may be beneficial to develop a plugin and reuse the code which otherwise would have to be maintained for each request separately.

What I am trying to say (without knowing the complexity behind) is that the possibility to jump to an arbitrary request in a collection/folder would make Insomnia shine in the area of similar tools. But I understand that the product roadmap is a result of many factors of which this proposal may be irrelevant.

Once again, thank you.

Cheers,
Jan
